### PR TITLE
[JENKINS-71732] Add component to collect executors information

### PR DESCRIPTION
**Pull request by Dohbedoh**
[JENKINS-71732](https://issues.jenkins.io/browse/JENKINS-71732): when recently troubleshooting issue with pipeline execution leaks on heavyweight executors (https://issues.jenkins.io/browse/JENKINS-71692), I found it hard to get information about the state of the executor and what was running on it from a Support Bundle. And other components such as thread dumps / pipeline thread dumps did not help either... So proposing a component that dumps the node executors details (not selected by default).

### Testing done

* Generate a Support Bundle with the "Node Executors" component selected:

```
Nodes Executors
===========

  * built-in (Jenkins)
      - Executor #-1
          - active: true
          - busy: true
          - causesOfInterruption: []
          - currentWorkUnit: hudson.model.queue.WorkUnit@5497813c[work=item_visible_omission » item_national_mosaic » item_first_difficulty #26]
          - elapsedTime: 93413
          - executable: org.jenkinsci.plugins.workflow.job.WorkflowJob@74a96fcf[item_visible_omission/item_national_mosaic/item_first_difficulty]
          - idle: false
          - idleStartMilliseconds: 1690860144069
          - progress: 99
          - state: TERMINATED
  * computer_appropriate_vein
      - Executor #0
          - active: true
          - busy: true
          - causesOfInterruption: []
          - currentWorkUnit: hudson.model.queue.WorkUnit@23dd8d[work=item_visible_omission » item_national_mosaic » item_first_difficulty #26 (Sleep)]
          - elapsedTime: 80003
          - executable: ExecutorStepExecution.PlaceholderTask{runId=item_visible_omission/item_national_mosaic/item_first_difficulty#26,label=computer_appropriate_vein,context=CpsStepContext[5:node]:Owner[item_visible_omission/item_national_mosaic/item_first_difficulty/26:item_visible_omission/item_national_mosaic/item_first_difficulty #26],cookie=e3b4dd75-68bf-4a94-a9a7-77c98de899d6,auth=null}
          - idle: false
          - idleStartMilliseconds: 1690860144075
          - progress: 99
          - state: TERMINATED
```

* Start a pipeline that sleeps, then generate an Agent Bundle with the "Node Executors" component selected. validate the information:

```
Node Executors
===========

  * computer_interesting_liability
      - Executor #0
          - active: true
          - busy: true
          - causesOfInterruption: []
          - currentWorkUnit: hudson.model.queue.WorkUnit@511d2729[work=item_visible_omission » item_national_mosaic » item_first_difficulty #25 (Sleep)]
          - elapsedTime: 5299629
          - executable: PlaceholderExecutable:ExecutorStepExecution.PlaceholderTask{runId=item_visible_omission/item_national_mosaic/item_first_difficulty#25,label=computer_interesting_liability,context=CpsStepContext[5:node]:Owner[item_visible_omission/item_national_mosaic/item_first_difficulty/25:item_visible_omission/item_national_mosaic/item_first_difficulty #25],cookie=f957cc44-539e-488d-93c8-d323477f4d6a,auth=null}
          - idle: false
          - idleStartMilliseconds: 1690857010516
          - progress: 99
          - state: TERMINATED
```

* Start a pipeline that sleeps, then abort it and immediately generate an Agent Bundle with the "Node Executors" component selected. Validate the information, including cause of interruptions in that case:

```
Node Executors
===========

  * computer_appropriate_vein
      - Executor #0
          - active: true
          - busy: true
          - causesOfInterruption: [Aborted by user_terminal_composer]
          - currentWorkUnit: hudson.model.queue.WorkUnit@23dd8d[work=part of item_visible_omission » item_national_mosaic » item_first_difficulty #26]
          - elapsedTime: 97796
          - executable: ExecutorStepExecution.PlaceholderTask{runId=item_visible_omission/item_national_mosaic/item_first_difficulty#26,label=computer_appropriate_vein,context=CpsStepContext[5:node]:Owner[item_visible_omission/item_national_mosaic/item_first_difficulty/26:item_visible_omission/item_national_mosaic/item_first_difficulty #26],cookie=e3b4dd75-68bf-4a94-a9a7-77c98de899d6,auth=null}
          - idle: false
          - idleStartMilliseconds: 1690860161868
          - progress: 99
          - state: TERMINATED
```

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    return "Node Executors";
+  }
+
+  @Override
+  public void addContents(@NonNull Container container) {
+    container.add(new PrefilteredPrintedContent("executors.md") {
+        @Override
+        public void printTo(PrintWriter out, ContentFilter filter) {
+          try {
+            out.println("Nodes Executors");
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java:79]:
```suggestion
            out.println("Node Executors");
```
or
```suggestion
            out.println("Executors of Nodes");
```

File: src/test/java/com/cloudbees/jenkins/support/impl/NodeExecutorsTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Test
+    public void addContents() throws Exception {
+        DumbSlave agent = j.createOnlineSlave(Label.parseExpression("test"), null);
+
+        Map<String, String> output = SupportTestUtils.invokeComponentToMap(new NodeExecutors(), agent.toComputer());
+
+        assertTrue(output.keySet().stream().anyMatch(key -> key.matches("executors.md")));
```
[the remark this answers is not in the document]
Comment (Dohbedoh) [reply src/test/java/com/cloudbees/jenkins/support/impl/NodeExecutorsTest.java:29]: Will do.

File: src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        if (workUnit != null) {
+          out.println("          - currentWorkUnit: " + ContentFilter.filter(filter, workUnit.toString()));
+        }
+        out.println("          - elapsedTime: " + executor.getElapsedTime());
+        Queue.Executable executable = executor.getCurrentExecutable();
+        if (executable != null) {
+          out.println("          - executable: " + ContentFilter.filter(filter, executable.getParent().toString()));
+        }
+        out.println("          - idle: " + executor.isIdle());
+        out.println("          - idleStartMilliseconds: " + executor.getIdleStartMilliseconds());
```
[the remark this answers is not in the document]
Comment (duemir) [reply src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java:142]: Is that the one that takes locale into account? Bundles with Japanese dates were inconvenient to analyze. So maybe some other formating function that is locale independent.

File: src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                .map(CauseOfInterruption::getShortDescription)
+                .collect(Collectors.joining(","))) + "]");
+        WorkUnit workUnit = executor.getCurrentWorkUnit();
+        if (workUnit != null) {
+          out.println("          - currentWorkUnit: " + ContentFilter.filter(filter, workUnit.toString()));
+        }
+        out.println("          - elapsedTime: " + executor.getElapsedTime());
+        Queue.Executable executable = executor.getCurrentExecutable();
+        if (executable != null) {
+          out.println("          - executable: " + ContentFilter.filter(filter, executable.getParent().toString()));
```
[the remark this answers is not in the document]
Comment (Dohbedoh) [reply src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java:139]: Thanks for spotting this.
About `currentWorkUnit`, I noticed that sometimes it add more information (like the fact that the subtask is an `AfterRestartTask` for example). So could still be useful ?

File: src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dohbedoh) [reply src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java:142]: maybe the `XS_DATETIME_FORMATTER` then: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/Util.java#L1859

File: src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        public void printTo(PrintWriter out, ContentFilter filter) {
+          try {
+            out.println("Nodes Executors");
+            out.println("===========");
+            out.println();
+
+            out.println("  * built-in (Jenkins)");
+            dumpExecutorInfo(Jenkins.get().toComputer(), out, filter);
+
+            Jenkins.get().getNodes().forEach(node -> {
```
[the remark this answers is not in the document]
Comment (Dohbedoh) [reply src/main/java/com/cloudbees/jenkins/support/impl/NodeExecutors.java]: Will do.

**Review by jglick — APPROVED** [review]

**Pull request merged**